File: algorithm/Python3_version/leetcode1144.decrease-elements-to-make-array-zigzag.py
class Solution:
    def movesToMakeZigzag(self, nums: List[int]) -> int:
        op1, op2 = 0, 0
        n = len(nums)
        for i in range(n):
            cur = nums[i]
            if i&1:
                if cur >= nums[i-1]:
                    op1 += cur-nums[i-1]+1
                    cur = nums[i-1]-1
                if i+1 < n and cur >= nums[i+1]:
                    op1 += cur-nums[i+1]+1
            else:
                if i>0 and cur >= nums[i-1]:
                    op2 += cur-nums[i-1]+1
                    cur = nums[i-1]-1
                if i+1 < n and cur >= nums[i+1]:
                    op2 += cur-nums[i+1]+1
        return min(op1, op2)

File: algorithm/leetcode1144.decrease-elements-to-make-array-zigzag.cpp
class Solution {
public:
    int movesToMakeZigzag(vector<int>& nums) {
    	int i = 1, n = nums.size(), op1 = 0, op2 = 0;
    	int cur1;
    	for(i = 0; i < n; i++) 
    	{
    		//目标: 都小于两侧
    		if(i&1)//奇数
    		{
    			cur1 = nums[i];
	    		if(cur1 >= nums[i-1])
	    		{
	    			op1 += cur1-nums[i-1]+1;
	    			cur1 = nums[i-1]-1;
	    		}
	    		if(i+1 < n && cur1 >= nums[i+1])
	    		{
	    			op1 += cur1-nums[i+1]+1;
	    		}
    		}
    		else
    		{
    			cur1 = nums[i];
	    		if(n > 0 && cur1 >= nums[i-1])
	    		{
	    			op2 += cur1-nums[i-1]+1;
	    			cur1 = nums[i-1]-1;
	    		}
	    		if(i+1 < n && cur1 >= nums[i+1])
	    		{
	    			op2 += cur1-nums[i+1]+1;
	    		}
    		}
    	}
    	return min(op1, op2);
    }
};